File: python/108.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    """
        72 ms 70.67%
        14.8 MB 100.00%
    """
    def sortedArrayToBST(self, nums):
        length = len(nums)
        if length == 0:
            return None
        if length == 1:
            return TreeNode(nums[0])  # 多一个判断会提示速度
        
        mid = (length - 1) // 2
        tempNode = TreeNode(nums[mid])
        tempNode.left = self.sortedArrayToBST(nums[:mid])  # [0:mid] 比 [:mid]切片慢?
        tempNode.right = self.sortedArrayToBST(nums[mid+1:])
        return tempNode

File: cpp/108.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <vector>
using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return buildAVL(nums, 0, nums.size() - 1);
    }

    TreeNode* buildAVL(vector<int>& nums, int i, int j) {
    /*
     * 28 ms, faster than 51.03%
     * 25.1 MB, less than 16.22%
     * 
     */
        if (j < i)
            return nullptr;
        int mid = (j - i) / 2 + i;
        TreeNode* tempNode = new TreeNode(nums[mid]);
        tempNode->left = buildAVL(nums, i, mid - 1);
        tempNode->right = buildAVL(nums, mid + 1, j);
        return tempNode;
    }
};